File: 2023/day3/sol.py
#! /usr/bin/env python
import sys 


def part1(filename):
    score = 0
    total_lines = []
    non_symbols = {*"1234567890."}
    with open(filename, "r") as f:
        while (line := f.readline().strip()):
            total_lines.append(line)
    # [print(line) for line in total_lines]
    for idx, line in enumerate(total_lines):
        for jdx, char in enumerate(line):
            if char not in non_symbols:
                symbolidx = (idx, jdx)
                for i in range(idx-1, idx+2):
                    for j in range(jdx-1, jdx+2):
                        if total_lines[i][j].isdigit():
                            startpos, endpos = get_number_limits(total_lines[i], j)
                            part_number = int(total_lines[i][startpos:endpos])
                            newline = ''.join('.' if startpos<=index<endpos else
                                              c for
                                              index, c in
                                              enumerate(total_lines[i]))
                            total_lines[i] = newline
                            score += part_number
    print("Part 1 score: ", score)

def part2(filename):
    score = 0
    total_lines = []
    non_symbols = {*"1234567890."}
    with open(filename, "r") as f:
        while (line := f.readline().strip()):
            total_lines.append(line)
    # [print(line) for line in total_lines]
    for idx, line in enumerate(total_lines):
        for jdx, char in enumerate(line):
            if char == "*":
                n_found = 0
                gear_ratio =1 
                symbolidx = (idx, jdx)
                for i in range(idx-1, idx+2):
                    startend = set()
                    for j in range(jdx-1, jdx+2):
                        if total_lines[i][j].isdigit():
                            startpos, endpos = get_number_limits(total_lines[i], j)
                            if (startpos, endpos) not in startend:
                                startend.add((startpos, endpos))
                                n_found+= 1
                                part_number = int(total_lines[i][startpos:endpos])
                                gear_ratio *= part_number

                if n_found == 2:
                    score += gear_ratio
                if n_found > 2:
                    print("FUCK")
                    print("FUCK AT: ", idx+1, "col ", jdx+1)
                    exit(1)
                if n_found == 1:
                    print("Not a gear at lineno ", idx+1, "col ", jdx+1)

    print("Part 2 score: ", score)


def get_number_limits(linestr, j):
    backposition=j
    forwardposition = j
    maximum = len(linestr)
    while (linestr[backposition-1].isdigit()):
        backposition-=1
    while (forwardposition < maximum and linestr[forwardposition].isdigit()): # Remember the end is the limiter
        forwardposition+=1
    return backposition, forwardposition # (linestr[backposition:forwardposition])


def main():
    if len(sys.argv) == 2:
        part1(sys.argv[1])
        part2(sys.argv[1])
    else:
        print(f"Usage: {sys.argv[0]} filename")

if __name__ == '__main__':
    main()

File: 2022/day03/sol.c
#define __GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>

int get_index(char c);
void part1(char *filename);
void part2(char *filename);

int main(int argc, char *argv[]) {
  if (argc != 2) {
    printf("Usage: %s <filename>\n", argv[0]);
    exit(1);
  }
  part1(argv[1]);
  part2(argv[1]);
}

void part1(char *filename) {
  FILE *file = fopen(filename, "r");
  char *line = NULL;
  size_t len = 0;
  ssize_t nread;
  int score = 0;
  char c;
  int idx;
  int found[52] = {0};
  while ((nread = getline(&line, &len, file)) != -1) {
    for (int i = 0; i < nread / 2; ++i) {
      for (int j = nread / 2; j < nread; ++j) {
        if ((c = line[i]) == line[j]) {
          idx = get_index(c);
          found[idx] = 1;
        }
      }
    }
    for (int i = 0; i < 52; ++i) {
      score += (i + 1) * found[i];
      found[i] = 0;
    }
  }
  printf("Part 1: %d\n", score);
  free(line);
  fclose(file);
}

void part2(char *filename) {
  FILE *file = fopen(filename, "r");
  char *line[3] = {NULL};
  size_t len[3] = {0};
  ssize_t nread;
  int score = 0;
  char c;
  int idx;
  int found[52] = {0};
  int count = 0;
  while (getline(&(line[0]), &len[0], file) != -1) {
    len[0] = 0;
    if (getline(&(line[1]), &len[1], file)) {
      len[1] = 0;
      if (getline(&(line[2]), &len[2], file)) {
        len[2] = 0;
        while ((c = *(line[0])++) && c != '\n') {
          idx = get_index(c);
          found[idx] = 1;
        }
        while ((c = *(line[1])++) && c != '\n') {
          idx = get_index(c);
          if (found[idx] == 1) {
            found[idx] = 2;
          }
        }
        while ((c = *(line[2])++) && c != '\n') {
          idx = get_index(c);
          if (found[idx] == 2) {
            found[idx] = 3;
          }
        }
      }
    }
    for (int i = 0; i < 52; ++i) {
      score += (i + 1) * (int)(found[i] / 3);
      found[i] = 0;
    }
  }
  printf("Part 2: %d\n", score);
  fclose(file);
}


int get_index(char c){
  if ('A' <= c && c <= 'Z') {
    return c - 'A' + 26;
  } else if ('a' <= c && c <= 'z') {
    return c - 'a';
  }
  return -1;
}
